src/shim/crypto.rs

```diff
@@ -132,15 +135,48 @@ impl Signature {
         self.sig_type
     }
 
+    /// Authenticates the message signature using protocol-specific validation:
+    /// - Delegated: Uses the Ethereum message with RLP encoding for signature verification, Verifies message roundtrip integrity
+    /// - BLS/SECP: Standard signature verification
+    pub fn authenticate_msg(
+        &self,
+        eth_chain_id: EthChainId,
+        msg: &SignedMessage,
+        addr: &crate::shim::address::Address,
+    ) -> anyhow::Result<()> {
+        let mut sig = self.clone();
```

Comment:
Why do we need to clone the entire instance? Does it need to always be cloned?

Reply:
self is a `Signature` itself, cloning it, so we can update the signature bytes in `line 157`, without actually affecting the self.

Reply:
You can use this syntax as well, which is a bit more idiomatic too: https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax

Reply:
Let me re-phrase. Do you think cloning (a relatively costly operation; you need to heap-allocate the entire vector) is necessary in case the signature type is not `Delegated` (which is the case most of the time)?

Reply:
> ```rust
>  match self.sig_type {
> ```

yes, that's it, this will only clone in case of `Delegated` sig. and doing `*self` should be cheap here.

Reply:
@LesnyRumcajs Thanks for pointing it out, I think you're right it doesn't make sense at all. I will update it as per my last comment.